### ISSUE TITLE
Misleading error message on type mismatch

### ISSUE THREAD

**Issue by ppwwyyxx** [tensorflow/tensorflow#12454]
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Archlinux
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.3.0-rc2-20-g0787eee 1.3.0
- **Python version**: 3.6
```python
import tensorflow as tf
x = tf.get_variable('asdfds', shape=[10], dtype=tf.int32)
x*0.5
```
Prints:
```
Traceback (most recent call last):               
  File "a.py", line 9, in <module>               
    x*0.5                                        
TypeError: unsupported operand type(s) for *: 'Variable' and 'float' 
```
Variable with matching dtype can multiply with float, the problem here is int32 cannot multiply with float.
Since usually the dtype of some tensor is not directly written in code, this misleading message can cause confusions.

**Comment by ibrahimsharaf** [comment]
Hi @ppwwyyxx @ali01, I am new to open source, how can I start working on this?

**Comment by DjangoPeng** [comment]
@ppwwyyxx I think the error message is threw by Python interpreter directly. @ali01 Can we catch it and print the exacter error message? I do know the `Mul` op is registered on the kernels dir, but I failed to locate it. Could you give me some idea for fix the issue?

**Comment by ppwwyyxx** [comment]
Relevant code here:
https://github.com/tensorflow/tensorflow/blob/c9db2486ead95f000395af14919f0ae8f508e429/tensorflow/python/ops/math_ops.py#L894-L907

`x.__mul__(0.5)` returns NotImplement, then `0.5.__rmul__(x)` is called. Both failed therefore Python throws an error with its predefined message.

**Comment by ppwwyyxx** [comment]
Fixed by 2da8e8c60343d9e833f1c591c54ee2a27c240842
